libcodechecker/log/option_parser.py

```diff
@@ -308,218 +200,228 @@ def item(self):
+    # GCC flags.
+    # Where do we use this architecture during analysis and why?
+    if flag_iterator.item == '-arch':
+        target.append('-arch')
+        next(flag_iterator)
+        target.append(flag_iterator.item)
+        buildaction.target = flag_iterator.item
+        return True
+
+    return False
+
+
+def __get_language(flag_iterator, target, buildaction):
+    """
+    This function consumes -x flag which is followed by the language. This
+    language is then collected to the buildaction object and saved to the
+    target command.
+    """
+    # TODO: Known issue: a -x flag may precede all source files in the build
+    # command with different languages.
+    if flag_iterator.item == '-x':
+        target.append('-x')
+        next(flag_iterator)
+        target.append(flag_iterator.item)
+        buildaction.lang = flag_iterator.item
+        return True
+
+    return False
+
+
+def __get_output(flag_iterator, target, buildaction):
+    """
+    This function consumes -o flag which is followed by the output file of the
+    action. This file is then collected to the buildaction object and saved to
+    the target command.
+    """
+    if flag_iterator.item == '-o':
+        target.append('-o')
+        next(flag_iterator)
+        target.append(flag_iterator.item)
+        buildaction.output = flag_iterator.item
+        return True
+
+    return False
+
+
+def __replace(flag_iterator, target, buildaction):
+    """
+    This function extends the target list with the corresponding replacement
+    based on REPLACE_OPTIONS_MAP if the flag_iterator is currently pointing to
+    a flag to replace.
+    """
+    value = REPLACE_OPTIONS_MAP.get(flag_iterator.item)
+
+    if value:
+        target.extend(value)
+
+    return bool(value)
+
+
+def __skip(flag_iterator, target, buildaction):
+    """
+    This function skips the flag pointed by the given flag_iterator with its
+    parameters if any.
+    """
+    if IGNORED_OPTIONS.match(flag_iterator.item):
+        return True
+
+    for pattern, arg_num in IGNORED_PARAM_OPTIONS.items():
+        if re.match(pattern, flag_iterator.item):
+            for i in range(arg_num):
+                next(flag_iterator)
+            return True
+
+    return False
+
+
+def parse_options(gcc_command):
+    """
+    This function parses a GCC compilation action and returns a BuildAction
+    object which can be the input of Clang analyzer tools.
+
+    gcc_command -- A valid GCC build command.
+    """
+
+    buildaction = BuildAction()
+
+    # Keep " characters.
+    gcc_command = gcc_command.replace(r'"', r'"\"')
+    gcc_command = shlex.split(gcc_command)
+
+    buildaction.action_type = buildaction.COMPILE
+    buildaction.compiler = gcc_command[0]
+    if '++' in buildaction.compiler or 'cpp' in buildaction.compiler:
+        buildaction.lang = 'c++'
+
+    clang_command = [gcc_command[0]]
+
+    flag_transformers = [
+        __get_arch,
```

Comment:
I think we should put transformers that are likely to be more common to the beginning and transformers that are likely to seldom fire to the end.